### PR TITLE
Resnet50/nn graph chengpeng

### PR DESCRIPTION
**Pull request by CPFLAME**
resnet50 net graph run

### 这个pr要解决的：
- [x] 搭Module模型，复用之前的
- [x] 单卡训练执行train
- [x] 单卡验证执行eval
- [x] loss 对齐，单卡graph  和 eager 
- [x] 添加 infer 脚本
- [x] 把 resnet50 的dataloader改成nn.Module风格
- [x] 把resnet50里的dataloader Module加入nn.Graph中

### CODE REVIEW

**Comment by Ldpe2G** [comment]
## nn.Graph验证结果
- 模型：resnet50
- 数据集：imagenette
- epochs：10
- optimizer：SGD
- learning rate：0.001
- momentum：0.9
- batchsize（train/val）：16/16

## Graph vs Eager
### Training Losses
graph_eager_loss

### Training abs losses
比较Eager和Graph两种模式下Loss的差异（取绝对值）
graph_eager_abs_loss

### Validation Accuracy
graph_eager_acc

### training epoch time
graph_eager_train_epoch_time

### training iter time
graph_eager_train_iter_time

### Evaluation Time (Epochs)
graph_eager_epoch_val_time

File: resnet50/check/check.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                eager_loss = criterion(logits, label)
+                eager_loss.backward()
+                eager_optimizer.step()
+                eager_optimizer.zero_grad()
+                eager_loss.numpy()
+                eager_iter_end_time = time.time()
+
+                # print("=====graph fc param=====")
+                # print(model_train_graph.graph_model.state_dict()['fc.bias']._origin[:50])
+                # print("=====eager fc param=====")
```
Comment (CPFLAME) [inline resnet50/check/check.py:182]: 此处可以观察训练时，两个model 的参数是否一致。

- eager与graph模型训练时，优化后model参数不一致问题
在graph和eager模式下训练，两者的model初始化参数一样，每次送进去的图片也是同一个dataloader的，optim都是用的SGD，loss也是同类型的，  但是经过几轮迭代后，两者的model 参数会有细微的差别，大概在1e-3~1e-4的量级

image

File: resnet50/check/check.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+            for b in range(len(train_data_loader)):
+                image, label = train_data_loader()
+                image = image.to("cuda")
+                label = label.to("cuda")
+
+                # oneflow graph train
+                graph_iter_start_time = time.time()
+                graph_loss = model_train_graph(image, label)
+                graph_loss.numpy()
```
Comment (CPFLAME) [inline resnet50/check/check.py:167]: 此处为了获取准确运行时间，加入.numpy()操作，用来进行CPU于GPU的同步

File: resnet50/check/check.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+                graph_iter_end_time = time.time()
+
```
Comment (CPFLAME) [inline resnet50/check/check.py:169]: 此处为了获取准确运行时间，加入.numpy()操作，用来进行CPU于GPU的同步

File: resnet50/check/check.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                eager_optimizer.zero_grad()
+                eager_loss.numpy()
+                eager_iter_end_time = time.time()
+
+                # print("=====graph fc param=====")
+                # print(model_train_graph.graph_model.state_dict()['fc.bias']._origin[:50])
+                # print("=====eager fc param=====")
+                # print(eager_model.state_dict()['fc.bias'][:50])
+                # import pdb
+                # pdb.set_trace()
```
Comment (CPFLAME) [inline resnet50/check/check.py:185]: 此处可以观察训练时，两个model 的参数是否一致。

- eager与graph模型训练时，优化后model参数不一致问题
在graph和eager模式下训练，两者的model初始化参数一样，每次送进去的图片也是同一个dataloader的，optim都是用的SGD，loss也是同类型的，  但是经过几轮迭代后，两者的model 参数会有细微的差别，大概在1e-3~1e-4的量级

image

File: resnet50/graph/train.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+python3 graph/train.py \
+    --save_checkpoint_path $CHECKPOINT_PATH \
+    --ofrecord_path $OFRECORD_PATH \
+    --learning_rate $LEARNING_RATE \
+    --mom $MOM \
+    --epochs $EPOCH \
+    --train_batch_size $TRAIN_BATCH_SIZE \
+    --val_batch_size $VAL_BATCH_SIZE \
+    #--load_checkpoint $LOAD_CHECKPOINT
```
Comment (rentainhe) [inline resnet50/graph/train.sh:29]: 注释部分可以删除~，然后make sure一下每个train.sh文件的超参是一致的，还有check.sh的超参也和train.sh超参一致，然后infer.sh也要一致

File: resnet50/check/check.py (hunk 4 above)
Comment (rentainhe) [reply resnet50/check/check.py:185]: 这边感觉可以记录一下grad绝对值差异的平均值，然后可以加一个trainer.avg_grad_diff，然后这样可以把注释去掉

File: resnet50/check/check.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    )
+    graph_optimizer = flow.optim.SGD(
+        graph_model.parameters(), lr=args.learning_rate, momentum=args.mom
+    )
+
+    # criterion setup
+    criterion = flow.nn.CrossEntropyLoss()
+    criterion = criterion.to("cuda")
+
+    class ModelTrainGraph(flow.nn.Graph):
```
Comment (rentainhe) [inline resnet50/check/check.py:79]: 这里可以尝试将dataloader也放到Graph里，可以参考一下AlexNet里的做法

File: resnet50/check/check.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+
+if __name__ == "__main__":
+    args = _parse_args()
+    trainer = Trainer(args)
+    compare_dic = setup(args)
+    print("init done")
+    trainer.compare_eager_graph(compare_dic)
+    del compare_dic
+    trainer.save_result()
```
Comment (rentainhe) [inline resnet50/check/check.py:323]: 可以像AlexNet里一样把trainer里的存的一些作图数据保存下来

File: resnet50/check/check.py [hunk 8]
```diff
@@ -177,12 +179,8 @@ def compare_eager_graph(self, compare_dic):
...
                 eager_iter_end_time = time.time()
 
-                # print("=====graph fc param=====")
-                # print(model_train_graph.graph_model.state_dict()['fc.bias']._origin[:50])
-                # print("=====eager fc param=====")
-                # print(eager_model.state_dict()['fc.bias'][:50])
-                # import pdb
-                # pdb.set_trace()
+                model_param_diff = compare_model_params(eager_model, model_train_graph)
+                self.eager_graph_model_diff_list.append(model_param_diff)
```
Comment (CPFLAME) [inline resnet50/check/check.py:183]: 在每次迭代后比较两者model的参数差异，得到均值

File: resnet50/check/check.py [hunk 9]
```diff
@@ -281,12 +282,52 @@ def save_result(self, ):
...
+        
+        save_results(self.graph_train_epoch_time_list, os.path.join(training_results_path, 'graph_train_epoch_time_list.txt'))
+        save_results(self.eager_train_epoch_time_list, os.path.join(training_results_path, 'eager_train_epoch_time_list.txt'))
+        
+        save_results(self.graph_eval_epoch_time_list, os.path.join(training_results_path, 'graph_eval_epoch_time_list.txt'))
+        save_results(self.eager_eval_epoch_time_list, os.path.join(training_results_path, 'eager_eval_epoch_time_list.txt'))
+        
+        save_results(self.eager_graph_model_diff_list, os.path.join(training_results_path, 'eager_graph_model_diff_list.txt'))
+
+        print("Results saved to: ", training_results_path)
```
Comment (CPFLAME) [inline resnet50/check/check.py:314]: 添加 画图的结果文件

File: resnet50/check/check.sh [hunk 10]
```diff
@@ -13,7 +13,7 @@ fi
 
 LEARNING_RATE=0.001
 MOM=0.9
-EPOCH=10
+EPOCH=20
```
Comment (CPFLAME) [inline resnet50/check/check.sh:16]: 统一epoch参数

File: resnet50/eager/train.sh [hunk 11]
```diff
@@ -13,7 +13,7 @@ fi
 
 LEARNING_RATE=0.001
 MOM=0.9
-EPOCH=90
+EPOCH=20
```
Comment (CPFLAME) [inline resnet50/eager/train.sh:16]: 统一epoch参数

File: resnet50/check/check.py (hunk 3 above)
Comment (strint) [reply resnet50/check/check.py:171]: 这个注释加到脚本中，这样外部开发者也可以看到，学习到这个处理技巧，而不会认为这里是无用的代码

File: resnet50/check/check.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+            for b in range(len(train_data_loader)):
+                image, label = train_data_loader()
+                image = image.to("cuda")
+                label = label.to("cuda")
+
+                # oneflow graph train
+                graph_iter_start_time = time.time()
+                graph_loss = model_train_graph(image, label)
+                graph_loss.numpy() # for synchronize CPU and GPU, get accurate running time
+                graph_iter_end_time = time.time()
```
Comment (CPFLAME) [inline resnet50/check/check.py:170]: 此处为了获取准确运行时间，加入.numpy()操作，用来进行CPU于GPU的同步， 已加入注释

File: resnet50/check/check.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+
+if __name__ == "__main__":
+    args = _parse_args()
+    trainer = Trainer(args)
+    compare_dic = setup(args)
+    print("init done")
+    trainer.compare_eager_graph(compare_dic)
+    del compare_dic
+    trainer.save_result()
```
Comment (rentainhe) [inline resnet50/check/check.py:364]: 添加一行空行，363行后

File: resnet50/graph/infer.sh [hunk 14]
```diff
@@ -0,0 +1,18 @@
+set -aux
+
+PRETRAIN_MODEL_PATH="resnet50_imagenet_pretrain_model"
+IMAGE_PATH="data/fish.jpg"
+# IMAGE_PATH="data/tiger.jpg"
```
Comment (rentainhe) [inline resnet50/graph/infer.sh:5]: infer这里的注释可以删除

Comment (CPFLAME) [reply resnet50/graph/infer.sh:5]: ok

**Review by rentainhe — APPROVED** [review]
